docs/interfaces.md

```diff
@@ -0,0 +1,431 @@
+
+// creates a new channel
+public int ecall_join_channel(
+        [in, size=gen_len] uint8_t *genesis, uint32_t gen_len);
+
+// pushes next block to tlcc
+public int ecall_next_block(
+        [user_check] uint8_t *block_bytes, uint32_t block_size);
+
+// verify state data
+public int ecall_get_state_metadata(
+        [in, string] const char *key, // TBD: should we add namespace separetely?
+        [in, size=32] uint8_t *nonce,
+        [out] sgx_cmac_128bit_tag_t *cmac);
+
+public int ecall_get_multi_state_metadata(
+        [in, string] const char *comp_key,
+        [in, size=32] uint8_t *nonce,
+        [out] sgx_cmac_128bit_tag_t *cmac);
+
+// verify msp identities
+public int ecall_validate_msp_id(
+        [in, string] const char *msp_id,
+        [in, size=32] uint8_t *nonce,
+        [out] bool isValid, 
+        [out] sgx_cmac_128bit_tag_t *cmac);
+```
+
+## State:
+The internal TLCC state consists of the current ledger state (block number, integrity-metadata, and msp) and sessions information for ecc communication.
+
+```c++
+// TODO: this needs some more brain cycles with respect of #402 and #410
+typedef struct internal_tlcc_state_t {
+        const char* INTEL_PUB_PEM, // Intel IAS verification key
+        
+        uint32_t sequence_number, // block number counter
+
+        // integrity metadata
+        kvs_t metadata_state,
+
+        // msp 
+        X509_STORE* root_certs_orderer, // root cert store for orderer org
+        X509_STORE* root_certs_apps, // root cert store for application orgs
+
+        // tlcc session
+        session_keys_t session_keys,
+};
+
+typedef struct kvs_t {};
+
+typedef struct session_keys_t {};
+```
+
+
+# ECC
+
+## Interface:
+The ECC interface is implemented through a "normal" chaincode interface. Methods with leading underscores
+are treated as FPC commands. Normal `invoke` invocations are forwarded to a FPC chaincode enclave.
+
+```go
+// chaincode interface (exposed to admin/clients) implemented by invoke
+func Invoke(stub shim.ChaincodeStubInterface) pb.Response {}
+
+// triggered by an admin
+CreateEnclave(params CC_Parameters) (Credentials, error)
+
+// key generatation 
+GenerateCCKeys(params CC_Parameters) (CCKey_registration_message, error)
+ExportCCKeys(credentals Credentials) (ExportMessage, error)
+ImportCCKeys(params CC_Parameters) (CCKey_registration_message, error)
+
+// chaincode invoke
+ChaincodeInvoke(transactionProposal []byte, ctx interface{}) (response, error)
+```
+
+```go
+// ECC must trigger this when sealedState is available using load() and ecc starts; this is not really a public interface
+createEnclaveWithSealedState(sealedState []byte) error
+
+// Storage interface
+// TODO how to store sealedState, where and under what identifier (e.g. chaincode_id);
+load(chaincode_id string) (sealedState []byte, err error)
+store(chaincode_id string, sealedState []byte) error
+```
+
+This interface is implemented by ECC to let a chaincode enclave call into the peer
```

Comment:
I think this is ideally outsourced into a separate sealedstorage module where you just have store and retrieve function exposed to ecc_enclave.  (with the enclave having some state-machine { no-keys, enclave-keys, cc-keys} and would make sure that certain function can be called only in a certain state and if not yet in that state, will try first to retrieve the state from sealed storage?)

Reply:
just in case it was not clear. i think it would be useful to explictly add the state-machine i mentioned above. Yes, as other state it doesn't really have to be fully-specified but something like this will be a crucial state component?

Reply:
ack! ok added a FSM state to the ecc state. Still need to define possible states.

Reply:
what about the states i have above as starters ;-)